src/js/tjs/ffi.js

```diff
@@ -0,0 +1,716 @@
+		super(ffitype, {
+			toBuffer: (obj, ctx)=>{
+				const buf = new Uint8Array(this._ffiType.size);
+				const offsets = this._ffiType.offsets;
+				for(let i=0; i<offsets.length; i++){
+					const [field, type] = this._fields[i];
+					if(obj.hasOwnProperty(field)){
+						let sbuf = type.toBuffer(obj[field], ctx);
+						buf.set(sbuf, offsets[i]);
+					}
+				}
+				return buf;
+			}, 
+			fromBuffer: (buf, ctx)=>{
+				let obj = {};
+				const offsets = this._ffiType.offsets;
+				for(let i=0; i<offsets.length; i++){
+					const [field, type] = this._fields[i];
+					const fbuf = buf.slice(offsets[i], offsets[i] + type.size);
+					obj[field] = type.fromBuffer(fbuf, ctx);
+				}
+				return obj;
+			}, 
+			name
+		});
+		this._fields = fields;
+	}
+	get fields(){
+		return this._fields;
+	}
+}
+
+export class ArrayType extends AdvancedType{
+	constructor(type, length, name){
+		const ffitype = type.ffiType ? type.ffiType : type;
+		const ffisz = ffitype.size;
+		super(ffitype, {
+			toBuffer: (arr, ctx)=>{
+				if(arr.length > this._length)
+					throw new RangeError('Array length exceeds type length');
+				const buf = new Uint8Array(ffisz*length);
+				for(let i=0; i<arr.length; i++){
+					let sbuf = type.fromBuffer(arr[i], ctx);
+					buf.set(sbuf, i*ffisz);
+				}
+				return buf;
+			}, 
+			fromBuffer: (buf, ctx)=>{
+				let arr = [];
+				for(let i=0; i<this._length; i++){
+					arr[i] = type.fromBuffer(buf.slice(i*ffisz, (i+1)*ffisz), ctx);
+				}
+				return arr;
+			}, 
+			name,
+		});
+		this._type = type;
+		this._length = length;
+	}
+	get ffiTypeStruct(){
+		if(!this._ffiStruct)
+			this._ffiStruct = new ffiInt.FfiType(this._length, this._ffiType);
+		return this._ffiStruct;
+	}
+	get length(){
+		return this._length;
+	}
+	get size(){
+		return this._ffiType.size * this._length;
+	}
+}
+
+export class StaticStringType extends ArrayType{
+	constructor(length, name){
+		super(types.sint8, length, name);
+	}
+	toBuffer(str, ctx = {}){
+		const txtBuf = (new TextEncoder()).encode(str);
+		return super.toBuffer(txtBuf, ctx);
+	}
+	fromBuffer(buf, ctx = {}){
+		return ffiInt.getCString(ffiInt.getArrayBufPtr(buf), buf.length);
+	}
+}
+
+export function errno(){
+	return ffiInt.errno();
+}
+
+export function strerror(err = errno()){
+	return ffiInt.strerror(err);
+}
+
+
+export class JSCallback{
+	constructor(rtype, argtypes, func){
+		this._func = (...args) => {
+			const arr = [];
+			const ctx = {};
+			for(let i=0;i<argtypes.length;i++){
+				ctx[i] = {};
+				arr.push(argtypes[i].fromBuffer(args[i], ctx[i]));
+			}
+			const ret = func(...arr);
+			return rtype.toBuffer(ret);
+		};
+		this._rtype = rtype;
+		this._argtypes = argtypes;
+		this._cif = new ffiInt.FfiCif(rtype.ffiType ?? rtype, ...argtypes.map(t => t.ffiType ?? t));
+		this._closure = new ffiInt.FfiClosure(this._cif, this._func);
+	}
+	get addr(){
+		return this._closure.addr;
+	}
+}
+
+// export only for test, not for user
+export function _parseCProto(header){
```

Comment:
Does this rely on anything not available to userland code? If not, can we move it to some ffiutils.js helper file for the tests?

Reply:
Not exactly sure how you meant it, look at the new commits.